[FSDP] Add `ignored_modules` ctor arg

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#75431 [FSDP] Add `ignored_modules` ctor arg**

https://github.com/pytorch/pytorch/issues/75255

Differential Revision: [D35550144](https://our.internmc.facebook.com/intern/diff/D35550144)

Commit:
[FSDP] Add `ignored_modules` ctor arg

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -340,8 +363,8 @@ def __init__(
         # sharding individual layers of a Module, with an outer wrapper to
         # shard any leftover parameters.
         params = []
-        for param_name, param in module.named_parameters():
```
Simplifying since `param_name` is unused

Commit:
Update on "[FSDP] Add `ignored_modules` ctor arg"


https://github.com/pytorch/pytorch/issues/75255

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -279,6 +279,8 @@ class FullyShardedDataParallel(nn.Module):
             algorithms to help backward communication and computation overlapping.
             Pros and cons of each algorithm is explained in the class ``BackwardPrefetch``.
+        ignored_modules (Optional[Iterable[torch.nn.Module]]): Modules whose
+            parameters are ignored by this instance. (Default: ``None``)
```
Maybe add some details on what it means to be ignored (i.e. not sharded, not managed by FSDP).

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -289,9 +291,27 @@ def __init__(
+        self._ignored_param_to_param_name = {}
+        for param in ignored_params:
+            unflat_param_names = param_to_unflat_param_names[param]
+            if len(unflat_param_names) != 1:
```
Can we just simplify this to `if len != 1: raise ValueError(...)`?

awgu replied:
The `assert` is meant as an internal check. If it is violated, then there may be a case for which `_get_param_to_unflat_param_names()` is incorrect. It might be good to keep the `assert`.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -418,6 +440,29 @@ def _init_reshard_after_forward(self):
+    ) -> Set[torch.nn.Module]:
+        """
+        Checks that ``_ignored_modules`` is an iterable of
+        ``torch.nn.Module`` s.
```
and it returns a set of them?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -437,25 +482,28 @@ def check_is_root(self) -> bool:
+            submodule for submodule in module.modules()
+            if isinstance(submodule, FullyShardedDataParallel) and
+            (not root_only or submodule.check_is_root())
+        ]
```
is this a stylistic refactor?

awgu replied:
Yup, my bad on this. I added something logically at one point and reformatted but then removed the addition.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1784,7 +1838,11 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
+                    p not in self._ignored_param_to_param_name
+                    and p.requires_grad
+                    for p in m.parameters()
+                ):
```
maybe Update the conditions described in L1855 as well?

rohan-varma commented on torch/distributed/fsdp/wrap.py:
```diff
@@ -181,20 +198,28 @@ def _recursive_wrap(
     assert wrapper_cls is not None, "Must specify wrapper_cls"
     # Make sure no child is already wrapped.
     for _, child in module.named_modules():
+        if child in ignored_modules:
```
I think intuitively ignoring a module means that all its children should be ignored, just want to make sure these changes adhere to this?

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -415,6 +415,46 @@ def test_auto_wrap_smoke_test(self, fsdp_init_mode, cpu_offload):
+        self.assertTrue(isinstance(model.module[0], FSDP))
+        self.assertTrue(isinstance(model.module[1], nn.Linear))
+        self.assertTrue(isinstance(model.module[2], FSDP))
+        self.assertTrue(isinstance(model.module[2].module[0], nn.Linear))
```
just confirming these are the ignored modules?

awgu replied:
Yes. `model.module[1]` is `sequential[1]`, and `model.module[2].module[0]` is `sequential[2][0]`, which are ignored on L420.

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -321,6 +325,34 @@ def _create_module():
+            model.outer.bias: "outer.bias", model.outer.weight: "outer.weight",
+        }
+        fsdp_model = FSDP(model, ignored_modules=ignored_modules)
+        with FSDP.state_dict_type(fsdp_model, StateDictType.FULL_STATE_DICT):
```
nit: unneeded because it is the default

awgu replied:
Oh, turns out it is not default (anymore?).

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -321,6 +325,34 @@ def _create_module():
+        ignored_param_to_param_name = {
+            model.outer.bias: "outer.bias", model.outer.weight: "outer.weight",
+        }
+        fsdp_model = FSDP(model, ignored_modules=ignored_modules)
```
if `ignored_modules` is part of `model` that is already wrapped in FSDP, will they be ignored? Seems unlikely as these modules were probably already sharded by the first call (i.e. L331) and we can't "undo" that?

rohan-varma commented on test/distributed/fsdp/test_fsdp_ignored_modules.py:
```diff
@@ -0,0 +1,50 @@
+        nonignored_numel = total_numel - ignored_numel
+        with wrapped_model.summon_full_params():
+            flat_param_numel = wrapped_model.params[0].numel()
+            self.assertEqual(flat_param_numel, nonignored_numel)
```
Can we run a forward pass with ignored modules to ensure that everything is fine and we don't need additional changes to fwd/backward?

rohan-varma reviewed (commented):
Looks great, thanks for the quick turnaround! Have some comments inline.

Also, I think we need to ensure that we can support this with `enable_wrap`? i.e. the following should work and we should add test for it:

```
with enable_wrap(FSDP, ignored_modules=[...]):
    model = wrap(model) # model contains children that are in ignored_modules
```

Commit:
Update on "[FSDP] Add `ignored_modules` ctor arg"


https://github.com/pytorch/pytorch/issues/75255

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550144).

Commit:
Update on "[FSDP] Add `ignored_modules` ctor arg"


https://github.com/pytorch/pytorch/issues/75255

Differential Revision: [D35550144](https://our.internmc.facebook.com/intern/diff/D35550144)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550144).

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -465,6 +491,60 @@ def _init_reshard_after_forward(self):
+        ignored_modules = set(
+            child for module in ignored_root_modules
+            for child in module.modules()
+            if not isinstance(child, FullyShardedDataParallel) and
```
This shouldn't be an expected use case, do we absolutely need support for it do you think?

awgu replied:
Sounds good. I think we can disallow this. I will update the PR to error in this case.

awgu replied:
Discussed offline: will support this since not too much added complexity.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -465,6 +491,60 @@ def _init_reshard_after_forward(self):
+        """
+        return set(
+            p for m in ignored_modules for p in m.parameters()
+            if not isinstance(p, FlatParameter)
```
wondering how we can pass in a ignored_module that contains FlatParameter here?

awgu replied:
This happens when there is a module in `ignored_modules` that is not `FSDP` but contains a child module that is `FSDP`. This is tested in `test_ignored_modules_nested()` in `test_fsdp_ignored_modules.py`.

This is exactly the case we were debating whether or not to support. On one hand, we can simplify the implementation by disallowing *any* submodules in the modules passed into `ignored_modules` to be `FSDP`, or we can leave the current implementation that allows but skips `FSDP` submodules.

rohan-varma reviewed (approved):
LGTM! I think it's better to simplify the implementation such that we can only ignore modules whose `self` and all children are not FSDP, but what do you think?

Commit:
Update on "[FSDP] Add `ignored_modules` ctor arg"


https://github.com/pytorch/pytorch/issues/75255

Differential Revision: [D35550144](https://our.internmc.facebook.com/intern/diff/D35550144)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550144).

Commit:
Update on "[FSDP] Add `ignored_modules` ctor arg"


https://github.com/pytorch/pytorch/issues/75255

Differential Revision: [D35550144](https://our.internmc.facebook.com/intern/diff/D35550144)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `ignored_modules` ctor arg"


https://github.com/pytorch/pytorch/issues/75255

Differential Revision: [D35550144](https://our.internmc.facebook.com/intern/diff/D35550144)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `ignored_modules` ctor arg"


https://github.com/pytorch/pytorch/issues/75255

Differential Revision: [D35550144](https://our.internmc.facebook.com/intern/diff/D35550144)

[ghstack-poisoned]

zhaojuanmao reviewed (commented):
coming late for reviewing, but awesome work!
